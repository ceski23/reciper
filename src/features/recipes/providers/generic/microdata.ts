import { decodeHTML5Strict } from 'entities'
import { parse, toSeconds } from 'iso8601-duration'
import type { Recipe } from 'features/recipes/types'
import { isDefined } from 'lib/utils'
import { getTextFromNode } from 'lib/utils/dom'
import { getColorFromImage } from 'lib/utils/images'
import { parseValidNumber } from 'lib/utils/numbers'

const getMetaData = (element: Element | null) => {
	if (element === null) return undefined
	if (element.tagName === 'META') return element.getAttribute('content') ?? undefined

	return getTextFromNode(element)
}

const getTags = (root: Element) => {
	const tags = getMetaData(root.querySelector('[itemprop="keywords"]'))?.split(',')
	const category = getMetaData(root.querySelector('[itemprop="recipeCategory"]'))?.split(',')

	return [tags, category]
		.filter(isDefined)
		.flatMap(texts => texts.map(text => text.toLocaleLowerCase().trim()))
}

const getInstructions = (element: Element) =>
	Array
		.from(element.querySelectorAll('[itemprop="recipeInstructions"], [itemprop="itemListElement"][itemtype="http://schema.org/HowToStep"]'))
		.map(elem => {
			const text = elem.querySelector('[itemprop=text]')?.textContent?.trim() ?? elem.textContent?.trim()
			const image = Array.from(elem.querySelectorAll('[itemprop="image"]')).filter(elem =>
				// exclude elements with svg loaders fallbacks
				!elem.getAttribute('src')?.includes('data:image/svg')
			).at(0)?.getAttribute('src')
				?? elem.querySelector('[itemprop="image"]')?.getAttribute('data-src') ?? undefined

			return (text
				? ({
					text: decodeHTML5Strict(text),
					image,
				})
				: undefined)
		})
		.filter(isDefined)

const findMainImage = (doc: Document) => {
	const metaElement = doc.querySelector('meta[itemprop="image"]')

	if (metaElement) return metaElement.getAttribute('content') ?? undefined

	const imageElements = Array.from(doc.querySelectorAll('img[itemprop="image"]')).filter(elem =>
		// exclude elements with svg loaders fallbacks
		!elem.getAttribute('src')?.includes('data:image/svg')
	)

	return imageElements.at(0)?.getAttribute('src') ?? undefined
}

export const extractMicrodata = async (doc: Document) => {
	const root = doc.querySelector('[itemscope][itemtype="https://schema.org/Recipe"], [itemscope][itemtype="http://schema.org/Recipe"]')
	if (!root) throw Error('Couldn\'t find recipe')

	const nameElement = root.querySelector('[itemprop="name"]')
	const name = (nameElement?.tagName === 'META') ? nameElement.getAttribute('content') : nameElement?.textContent?.trim()

	const descriptionElement = root.querySelector('[itemprop="description"]')
	const description = descriptionElement ? getMetaData(descriptionElement)?.trim() : undefined

	const image = findMainImage(doc)

	const instructionSections = Array.from(root.querySelectorAll('[itemprop="recipeInstructions"][itemtype="http://schema.org/HowToSection"]'))
	const instructions: Recipe['instructions'] = instructionSections.length > 0
		? instructionSections.flatMap(sectionElement => {
			const sectionTitle = sectionElement.querySelector('[itemprop="name"]')?.getAttribute('content') ?? undefined

			return getInstructions(sectionElement).map(data => ({
				...data,
				group: sectionTitle,
			}))
		})
		: getInstructions(root)

	const prepTimeISO = root.querySelector('[itemprop="prepTime"]')?.getAttribute('content') ?? undefined
	const prepTime = prepTimeISO ? Math.round(toSeconds(parse(prepTimeISO)) / 60) : undefined

	const totalTimeISO = root.querySelector('[itemprop="totalTime"]')?.getAttribute('content') ?? undefined
	const totalTime = totalTimeISO ? Math.round(toSeconds(parse(totalTimeISO)) / 60) : undefined

	// Rating
	const ratingElement = root.querySelector('[itemprop="ratingValue"]')
	const rating = (ratingElement?.textContent)
		? Number.parseFloat(ratingElement.textContent)
		: undefined

	// Nutrition
	const caloriesElement = root.querySelector('[itemprop="calories"]')
	const calories = (caloriesElement?.textContent)
		? Number.parseInt(caloriesElement.textContent, 10)
		: undefined

	// TODO: keywords, nutrition, recipeCategory, recipeCuisine

	const ingredientsElements = root.querySelectorAll('[itemprop="recipeIngredient"]')
	const ingredients = Array
		.from(ingredientsElements)
		.map(elem => (elem.textContent ? ({ text: elem.textContent?.trim() }) : undefined))
		.filter(isDefined)

	const tags = Array.from(new Set(getTags(root)))

	const servingsElement = root.querySelector('[itemprop="recipeYield"]')
	const servings = parseValidNumber(getMetaData(servingsElement))

	const color = image ? await getColorFromImage(image.toString()) : undefined

	return {
		name: name ?? undefined,
		ingredients,
		description,
		image,
		prepTime: prepTime ?? totalTime,
		instructions,
		rating,
		calories,
		tags,
		servings,
		color,
	} satisfies Partial<Recipe>
}
