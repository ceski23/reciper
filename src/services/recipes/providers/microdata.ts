/* eslint-disable import/no-cycle */
import { RecipeScrapper } from '.';

// eslint-disable-next-line @typescript-eslint/require-await
const scrapper: RecipeScrapper = async (doc, url) => {
  const root = doc.querySelector('[itemscope][itemtype="https://schema.org/Recipe"], [itemscope][itemtype="http://schema.org/Recipe"]');
  if (!root) throw Error('Couldn\'t find recipe');

  const nameElement = root.querySelector('[itemprop="name"]');
  const name = (nameElement?.tagName === 'META') ? nameElement.getAttribute('content') : nameElement?.textContent;
  if (!name) throw Error('Couldn\'t find recipe name');

  const descriptionElement = root.querySelector('[itemprop="description"]');
  const description = ((descriptionElement?.tagName === 'META') ? descriptionElement.getAttribute('content') : descriptionElement?.textContent) ?? undefined;

  const imageElement = root.querySelector('[itemprop="image"]');
  const image = ((imageElement?.tagName === 'META') ? imageElement.getAttribute('content') : imageElement?.getAttribute('src')) ?? undefined;

  const instructionsElements = root.querySelectorAll('[itemprop="recipeInstructions"]');
  const instructions = Array
    .from(instructionsElements)
    .map((elem) => elem.textContent?.trim())
    .filter(Boolean) as string[];

  // if (instructions.length === 0) throw Error('No instructions found');

  const prepTime = root.querySelector('[itemprop="prepTime"]')?.getAttribute('content') || undefined;

  // Rating
  const ratingElement = root.querySelector('[itemprop="ratingValue"]');
  const rating = (ratingElement?.textContent)
    ? Number.parseFloat(ratingElement.textContent)
    : undefined;

  // Nutrition
  const caloriesElement = root.querySelector('[itemprop="calories"]');
  const calories = (caloriesElement?.textContent)
    ? Number.parseInt(caloriesElement.textContent, 10)
    : undefined;

  // TODO: keywords, nutrition, recipeCategory, recipeCuisine

  const ingredientsElements = root.querySelectorAll('[itemprop="recipeIngredient"]');
  const ingredients = Array
    .from(ingredientsElements)
    .map((elem) => elem.textContent?.trim())
    .filter(Boolean) as string[];

  // if (ingredients.length === 0) throw Error('No ingredients found');

  return {
    name,
    ingredients,
    description,
    image,
    prepTime,
    instructions,
    url,
    rating,
    calories,
    tags: [],
  };
};

export default scrapper;
