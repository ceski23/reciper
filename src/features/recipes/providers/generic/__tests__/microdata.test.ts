import { describe, expect, it } from 'vitest'
import { extractMicrodata } from '../microdata'
import { recipe1 } from './microdata.fixtures'

describe('should scrape recipes using Microdata format', () => {
	const scrapeRecipe = async (data: string) => extractMicrodata(new DOMParser().parseFromString(data, 'text/html'))

	it('should scrape valid recipe for Gofry bananowe', async () => {
		expect(await scrapeRecipe(recipe1)).toMatchSnapshot()
	})

	it('should not found recipe', async () => {
		await expect(scrapeRecipe(/* html */ `
          <div>
            <p>This is not recipe</p>
          </div>
        `)).rejects.toThrow('Couldn\'t find recipe')
	})

	it('should detect recipe\'s title', async () => {
		{
			const recipe = await scrapeRecipe(/* html */ `
                <div itemscope itemtype="http://schema.org/Recipe">
                  <h2 itemprop="name">Test recipe</h2>
                </div>
            `)

			expect(recipe.name).toBe('Test recipe')
		}
		{
			const recipe = await scrapeRecipe(/* html */ `
                <div itemscope itemtype="http://schema.org/Recipe">
                  <meta itemprop="name" content="Test recipe" />
                </div>
            `)

			expect(recipe.name).toBe('Test recipe')
		}
	})

	it('should detect recipe\'s image', async () => {
		{
			const recipe = await scrapeRecipe(/* html */ `
                <div itemscope itemtype="http://schema.org/Recipe">
                  <img itemprop="image" src="test.jpg" />
                </div>
            `)

			expect(recipe.image).toBe('test.jpg')
		}
		{
			const recipe = await scrapeRecipe(/* html */ `
                <div itemscope itemtype="http://schema.org/Recipe">
                  <meta itemprop="image" content="test.jpg" />
                </div>
            `)

			expect(recipe.image).toBe('test.jpg')
		}
	})

	it('should detect recipe\'s description', async () => {
		{
			const recipe = await scrapeRecipe(/* html */ `
                <div itemscope itemtype="http://schema.org/Recipe">
                  <p itemprop="description">Testing test</p>
                </div>
            `)

			expect(recipe.description).toBe('Testing test')
		}
		{
			const recipe = await scrapeRecipe(/* html */ `
                <div itemscope itemtype="http://schema.org/Recipe">
                  <meta itemprop="description" content="Testing test" />
                </div>
            `)

			expect(recipe.description).toBe('Testing test')
		}
	})

	it('should detect recipe\'s instructions', async () => {
		{
			const recipe = await scrapeRecipe(/* html */ `
                <div itemscope itemtype="http://schema.org/Recipe">
                  <span itemprop="recipeInstructions">Test test test</span>
                </div>
            `)

			expect(recipe.instructions).toEqual([
				{ text: 'Test test test' },
			])
		}
		{
			const recipe = await scrapeRecipe(/* html */ `
                <div itemscope itemtype="http://schema.org/Recipe">
                <span itemprop="recipeInstructions">Test1</span>
                <span itemprop="recipeInstructions">Test2</span>
                <span itemprop="recipeInstructions">Test3</span>
                </div>
            `)

			expect(recipe.instructions).toEqual([
				{ text: 'Test1' },
				{ text: 'Test2' },
				{ text: 'Test3' },
			])
		}
	})

	it('should detect recipe\'s preparation time', async () => {
		const recipe = await scrapeRecipe(/* html */ `
          <div itemscope itemtype="http://schema.org/Recipe">
            <meta itemprop="prepTime" content="PT10M" />
          </div>
        `)

		expect(recipe.prepTime).toBe(10)
	})

	it('should detect recipe\'s rating', async () => {
		const recipe = await scrapeRecipe(/* html */ `
          <div itemscope itemtype="http://schema.org/Recipe">
            <span itemprop="ratingValue">4.8</span>
          </div>
        `)

		expect(recipe.rating).toBe(4.8)
	})

	it('should detect recipe\'s calories', async () => {
		const recipe = await scrapeRecipe(/* html */ `
          <div itemscope itemtype="http://schema.org/Recipe">
            <span itemprop="calories">540 calories</span>
          </div>
        `)

		expect(recipe.calories).toBe(540)
	})

	it('should detect recipe\'s ingredients', async () => {
		const recipe = await scrapeRecipe(/* html */ `
          <div itemscope itemtype="http://schema.org/Recipe">
            <span itemprop="recipeIngredient">Test1</span>
            <span itemprop="recipeIngredient">Test2</span>
          </div>
        `)

		expect(recipe.ingredients).toStrictEqual([
			{ text: 'Test1' },
			{ text: 'Test2' },
		])
	})
})
