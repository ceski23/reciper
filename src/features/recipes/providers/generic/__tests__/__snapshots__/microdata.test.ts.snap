// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should scrape recipes using LD+JSON format > should scrape valid recipe for Gofry bananowe 1`] = `
{
  "calories": undefined,
  "color": "#ffffff",
  "description": "Gofry bananowe to idealny sposób na wykorzystanie
				dojrzałych bananów w swojej kuchni. Wyróżniają się prostotą przygotowania - wystarczy kilka składników i
				parę minut, aby przygotować pyszne i pożywne gofry, które zachwycą całą Twoją rodzinę.",
  "image": "https://poprostupycha.com.pl/wp-content/uploads/2021/06/gofry-bananowe-5-768x1151.jpg",
  "ingredients": [
    {
      "text": "1 duży, dojrzały banan (ewentualnie 2 małe)",
    },
    {
      "text": "190 g mąki pszennej",
    },
    {
      "text": "40 ml oleju",
    },
    {
      "text": "1 duże jajko",
    },
    {
      "text": "1 łyżeczka proszku do pieczenia",
    },
    {
      "text": "szczypta cynamonu",
    },
    {
      "text": "150 ml mleka",
    },
    {
      "text": "cukier puder do oprószenia",
    },
    {
      "text": "owoce do dekoracji",
    },
  ],
  "instructions": [
    {
      "image": "https://poprostupycha.com.pl/wp-content/uploads/2021/06/gofry-bananowe-7.jpg",
      "text": "Zaczynamy od przygotowania ciasta na
						gofry bananowe.
						Dojrzałego dużego banana (lub dwa małe) rozgniatamy np. za pomocą widelca lub tłuczka do
							ziemniaków.",
    },
    {
      "image": "https://poprostupycha.com.pl/wp-content/uploads/2021/06/gofry-bananowe-1.jpg",
      "text": "Dodajemy mąkę pszenną tortową (190 g),
						olej lub roztopione masło (40 ml), duże jajko, proszek do pieczenia (1 płaska łyżeczka), cynamon
						(szczypta) oraz mleko (150 ml).
						PS. Nie używam w tym przepisie cukru, bo użyłam słodkiego, dojrzałego banana, ale jeśli
							chcesz to dodaj 1 łyżkę cukru białego lub pudru.",
    },
    {
      "image": "https://poprostupycha.com.pl/wp-content/uploads/2021/06/gofry-bananowe-2.jpg",
      "text": "Całość mieszamy za pomocą rózgi
						kuchennej jedynie do momentu połączenia się składników. Ciasto ma wyjść nieco gęste.
						Odstawiamy je na 5-10 minut, aby się lekko napuszyło.",
    },
    {
      "image": "https://poprostupycha.com.pl/wp-content/uploads/2021/06/gofry-bananowe-3.jpg",
      "text": "Smażymy gofry w rozgrzanej gofrownicy
						aż będą ładnie rumiane.
						Podajemy z ulubionymi dodatkami. U mnie z serkiem homogenizowanym waniliowym i świeżymi
							owocami. Smacznego!",
    },
  ],
  "name": "poprostupycha.com.pl",
  "prepTime": 10,
  "rating": 4.6,
  "servings": 6,
  "tags": [
    "gofry",
  ],
}
`;

exports[`should scrape recipes using Microdata format > should scrape valid recipe for Gofry bananowe 1`] = `
{
  "calories": undefined,
  "color": "#ffffff",
  "description": "Gofry bananowe to idealny sposób na wykorzystanie
				dojrzałych bananów w swojej kuchni. Wyróżniają się prostotą przygotowania - wystarczy kilka składników i
				parę minut, aby przygotować pyszne i pożywne gofry, które zachwycą całą Twoją rodzinę.",
  "image": "https://poprostupycha.com.pl/wp-content/uploads/2021/06/gofry-bananowe-5-768x1151.jpg",
  "ingredients": [
    {
      "text": "1 duży, dojrzały banan (ewentualnie 2 małe)",
    },
    {
      "text": "190 g mąki pszennej",
    },
    {
      "text": "40 ml oleju",
    },
    {
      "text": "1 duże jajko",
    },
    {
      "text": "1 łyżeczka proszku do pieczenia",
    },
    {
      "text": "szczypta cynamonu",
    },
    {
      "text": "150 ml mleka",
    },
    {
      "text": "cukier puder do oprószenia",
    },
    {
      "text": "owoce do dekoracji",
    },
  ],
  "instructions": [
    {
      "image": "https://poprostupycha.com.pl/wp-content/uploads/2021/06/gofry-bananowe-7.jpg",
      "text": "Zaczynamy od przygotowania ciasta na
						gofry bananowe.
						Dojrzałego dużego banana (lub dwa małe) rozgniatamy np. za pomocą widelca lub tłuczka do
							ziemniaków.",
    },
    {
      "image": "https://poprostupycha.com.pl/wp-content/uploads/2021/06/gofry-bananowe-1.jpg",
      "text": "Dodajemy mąkę pszenną tortową (190 g),
						olej lub roztopione masło (40 ml), duże jajko, proszek do pieczenia (1 płaska łyżeczka), cynamon
						(szczypta) oraz mleko (150 ml).
						PS. Nie używam w tym przepisie cukru, bo użyłam słodkiego, dojrzałego banana, ale jeśli
							chcesz to dodaj 1 łyżkę cukru białego lub pudru.",
    },
    {
      "image": "https://poprostupycha.com.pl/wp-content/uploads/2021/06/gofry-bananowe-2.jpg",
      "text": "Całość mieszamy za pomocą rózgi
						kuchennej jedynie do momentu połączenia się składników. Ciasto ma wyjść nieco gęste.
						Odstawiamy je na 5-10 minut, aby się lekko napuszyło.",
    },
    {
      "image": "https://poprostupycha.com.pl/wp-content/uploads/2021/06/gofry-bananowe-3.jpg",
      "text": "Smażymy gofry w rozgrzanej gofrownicy
						aż będą ładnie rumiane.
						Podajemy z ulubionymi dodatkami. U mnie z serkiem homogenizowanym waniliowym i świeżymi
							owocami. Smacznego!",
    },
  ],
  "name": "poprostupycha.com.pl",
  "prepTime": 10,
  "rating": 4.6,
  "servings": 6,
  "tags": [
    "gofry",
  ],
}
`;
